File: Source/Aladdin.CAPI.GOST.Java/src/aladdin/capi/gost/mode/gostr3412/CTR.java
package aladdin.capi.gost.mode.gostr3412;
import aladdin.*; 
import aladdin.capi.*; 
import java.security.*; 
import java.io.*; 

///////////////////////////////////////////////////////////////////////////////
// Режим CTR
///////////////////////////////////////////////////////////////////////////////
public class CTR extends aladdin.capi.mode.CTR
{
    // режим смены ключа и размер смены ключа
    private final KeyDerive keyMeshing; private final int N; 
    
    // конструктор
	public CTR(Cipher engine, CipherMode.CTR parameters, KeyDerive keyMeshing, int N)
	{ 
        // сохранить переданные параметры
        super(engine, parameters); this.N = N; 
        
        // сохранить переданные параметры
        this.keyMeshing = RefObject.addRef(keyMeshing); 
	}
    // конструктор
	public CTR(Cipher engine, CipherMode.CTR parameters)
	{ 
        // сохранить переданные параметры
        super(engine, parameters); this.keyMeshing = null; N = 0;  
	}
    // деструктор
    @Override protected void onClose() throws IOException    
    { 
        // освободить ресурсы
        RefObject.release(keyMeshing); super.onClose();
    }
    // преобразование зашифрования
    @Override protected Transform createEncryption(ISecretKey key) 
    { 
        // преобразование зашифрования
        return new Transform(engine(), keyMeshing, N, key, mode()); 
    }
    // преобразование расшифрования
    @Override protected Transform createDecryption(ISecretKey key) 
    { 
        // преобразование расшифрования
        return new Transform(engine(), keyMeshing, N, key, mode()); 
    }
    ///////////////////////////////////////////////////////////////////////////////
    // Режим шифрования CFB
    ///////////////////////////////////////////////////////////////////////////////
    public static class Transform extends aladdin.capi.mode.CTR.Transform
    {
        // алгоритм смены ключа и размер смены ключа
        private final KeyDerive keyMeshing; private final int N;
        // текущий ключ и накопленный размер
        private ISecretKey currentKey; private int length; 

        // конструктор
        public Transform(Cipher engine, KeyDerive keyMeshing, int N, 
            ISecretKey key, CipherMode.CTR parameters) 
        { 
            // сохранить переданные параметры
            super(engine, key, parameters); this.N = N; 
            
            // проверить корректность параметров
            if ((N % engine.blockSize()) != 0) throw new IllegalArgumentException(); 
            
            // сохранить переданные параметры
            currentKey = RefObject.addRef(key); 
            
            // сохранить переданные параметры
            this.keyMeshing = RefObject.addRef(keyMeshing); 
        } 
        // конструктор
        public Transform(Cipher engine, ISecretKey key, CipherMode.CTR parameters) 
        { 
            // смена ключа отсутствует
            super(engine, key, parameters); keyMeshing = null; N = 0; 
            
            // сохранить переданные параметры
            currentKey = RefObject.addRef(key); 
        } 
        // освободить ресурсы
        @Override protected void onClose() throws IOException    
        {
            // освободить ресурсы
            RefObject.release(currentKey);
            
            // освободить ресурсы
            RefObject.release(keyMeshing); super.onClose();
        }
        // инициализировать алгоритм
        @Override public void init() throws IOException { super.init(); length = 0; }
        
        @Override protected void update( 
            byte[] data, int dataOff, byte[] buf, int bufOff) throws IOException
        {
            // обработать полный блок
            super.update(data, dataOff, buf, bufOff); 
            
            // увеличить размер данных
            length += blockSize(); if (N == 0 || (length % N) != 0) return; 

            // изменить значение ключа
            try (ISecretKey key = keyMeshing.deriveKey(currentKey, iv(), currentKey.keyFactory(), 32))
            {
                // переустановить ключ
                resetKey(key); RefObject.release(currentKey); currentKey = RefObject.addRef(key); 
            }
            // обработать неожидаемое исключение
            catch (InvalidKeyException e) { throw new RuntimeException(e); }
        }
    }
}

File: Source/Aladdin.CAPI.GOST/CAPI/GOST/Mode/GOSTR3412/CTR.cs
using System; 

namespace Aladdin.CAPI.GOST.Mode.GOSTR3412
{
    ///////////////////////////////////////////////////////////////////////////////
    // Режим CTR
    ///////////////////////////////////////////////////////////////////////////////
    public class CTR : CAPI.Mode.CTR
    {
        // режим смены ключа и размер смены ключа
        private KeyDerive keyMeshing; private int N; 
    
        // конструктор
	    public CTR(CAPI.Cipher engine, CipherMode.CTR parameters, 
            KeyDerive keyMeshing, int N) : base(engine, parameters)
	    { 
            // проверить корректность параметров
            if ((N % engine.BlockSize) != 0) throw new ArgumentException(); 

            // сохранить переданные параметры
            this.keyMeshing = RefObject.AddRef(keyMeshing); this.N = N; 
	    }
        // конструктор
	    public CTR(CAPI.Cipher engine, CipherMode.CTR parameters) 

            // сохранить переданные параметры
            : base(engine, parameters) { this.keyMeshing = null; N = 0; }

        // деструктор
        protected override void OnDispose() 
        { 
            // освободить ресурсы
            RefObject.Release(keyMeshing); base.OnDispose();
        }
        // преобразование зашифрования
        protected override CAPI.Transform CreateEncryption(ISecretKey key) 
        { 
            // преобразовать тип параметров
            CipherMode.CTR parameters = (CipherMode.CTR)Mode; 

            // преобразование расшифрования
            return new CTR_ENC(Engine, keyMeshing, N, key, parameters); 
        }
        // преобразование расшифрования
        protected override CAPI.Transform CreateDecryption(ISecretKey key) 
        { 
            // преобразовать тип параметров
            CipherMode.CTR parameters = (CipherMode.CTR)Mode; 

            // преобразование расшифрования
            return new CTR_ENC(Engine, keyMeshing, N, key, parameters); 
        }
    }
}